### ISSUE TITLE
compact block fingerprinting

### ISSUE THREAD

**Issue by Crypt-iQ** [bitcoin/bitcoin#28272]
I haven't written a test for this, but it seems that a peer can fingerprint a chunk of our mempool by announcing a new compact block with a valid header, but inserting many `shorttxids` that don't belong to the block but are suspected to be in our mempool. Assuming no short id collisions, we'll then request each tx that we didn't find in either `vExtraTxnForCompact` or the mempool. 

I'm not sure how to fix this since compact block relay relies on this behavior. One possible way of fixing this was brought up here https://github.com/bitcoin/bitcoin/pull/27086#issuecomment-1426937832.

**Comment by Crypt-iQ** [comment]
I think implementing the TODO might add round-trips and maybe leak what we had in our own mempool prior to block acceptance?

**Comment by Crypt-iQ** [comment]
There are two cases to think about here that currently leak some info:

1. we have the tx, so we won't request it
2. we don't have the tx, so we will request it

To solve both cases, we could choose some transactions to request even if we have them. For example, if we need X transactions to reconstruct the block, we could request Y transactions in GETBLOCKTXN where we have the other Y-X transactions. Adding these decoy transactions increases the BLOCKTXN response and could cause more round trips. We would also have to ignore the known transactions in BLOCKTXN.

If there are too many txn's that we're missing, we could fall back to block relay instead of trying to add many decoy transactions. According to [this data point](https://github.com/bitcoin/bitcoin/pull/27086#issuecomment-1426937832), compact block reconstruction works pretty well. Most of the time no reconstruction happens and sometimes only a few txn's are needed. So falling back to block relay should be unlikely.

Since quick compact block relay is pretty important to avoid stale blocks, I'm not really sure how to weigh the privacy-bandwidth tradeoff.

**Comment by mzumsande** [comment]
After #27675 (but also before, with a few more restrictions), we relay any tx from our mempool (except for those that arrived after the last INV sent to the peer), so someone wanting to fingerprint our mempool could just ask for those transactions directly instead I think and then check whether we send it to them or answer with `NOTFOUND`.

So I wonder why would someone trouble themselves with doing compact blocks for this, which would involve either creating a new valid block header (expensive) or trying to be the first to announce a compact block found by someone else to the victim (which has timing issues), if they could just ask for this info directly? (this doesn't apply to block-relay-only peers though).

**Comment by Crypt-iQ** [comment]
I meant to reply but guess I never did. The point about blocks-only nodes is a good one. In the happy path, they aren't sent compact blocks since they never send `SENDCMPCT` to the peer again to set the high-bandwidth mode. But they can still be sent unsolicited compact blocks and because they can't reconstruct the blocks, it lets them be fingerprinted if anything is in their mempool. In most cases I don't think we can send `GETDATA` to fetch txns from blocksonly nodes since they never `INV` us anything. Additionally, I think because the blocksonly nodes are unable to reconstruct the block, I think they can be stalled slightly from tip. The first connection can quickly announce a block to them via `HEADERS` and then a second connection can announce the block via `CMPCTBLOCK`. This increases the stalling timeout for the first connection by 5 minutes. This process can be repeated for the next block (but the one after that) because of this line:
https://github.com/bitcoin/bitcoin/blob/9da0820ec55e136d5213bf5bb90c36499debc034/src/net_processing.cpp#L4393

I can't see of a way to extend the block stalling, but to be safe it may just be wise to ignore unsolicited compact blocks in blocksonly mode.

**Comment by instagibbs** [comment]
Dropping seemingly-unsolicited compact blocks seems fine, we already do that for the parallel portion, this should be a change just to the first one? blocksonly nodes should never set cb peers.

**Comment by mzumsande** [comment]
Some discussion about the privacy implications for blocksonly peers in IRC: https://bitcoin-irc.chaincode.com/bitcoin-core-dev/2025-05-23#1123715;

Dropping unsolicted `cmpctblock` messages makes sense to me as well (and doesn't appear to be at odds with anything in BIP152).

**Comment by Crypt-iQ** [comment]
> So I had thought the code would check if there were too many missing transactions and just request a full block if all were missing. That doesn't address the general attack (since the attacker could admit some real txn that are in the block) but it would mostly address the blocksonly case (and could easily get a if blocksonly check so that it always did) -- but I can't find that. It should be added.

Is this still useful if we add a patch to drop unsolicited `cmpctblock` messages? An attacker would still be able to do this to non-blocksonly peers, but as pointed out above by @mzumsande it doesn't seem particularly useful as the attacker could simply request the transactions directly from these peers.

**Comment by davidgumberg** [comment]
Opened #32606 to drop unsolicited CMPCTBLOCK messages.

**Comment by Crypt-iQ** [comment]
> This was only half my point - my larger point was that we can't really fix this issue because if you're the fastest peer once you're gonna be the fastest peer again. I'm not sure it's worth trying to fix it broadly given the impact it might have (even if small) and the fact that it isn't really a great fix.

Hmm, I agree with this. I guess the solution is a better protocol to avoid fingerprinting? Adding more complexity here sounds a bit scary.

I looked at the WIP fibre patchset here: https://github.com/w0xlt/bitcoinfibre/blob/45897a826b37eb417cba93ac17ff7bebb272893c/src/net_processing.cpp#L873 and it seems like fibre does rely on sending `cmpctblock` even if the peer hasn't requested hb mode. I think this is how fibre bootstraps until it is marked as high bandwidth by the peer? In any case, my opinion was formed by reading the current code which does not do this and I don't think we should break fibre.

**Comment by davidgumberg** [comment]
> The relevant code is in UDPRelayBlock

`UDPRelayBlock` is relevant only for FIBRE to FIBRE connections, and these also have [bespoke](https://github.com/w0xlt/bitcoinfibre/blob/45897a826b37eb417cba93ac17ff7bebb272893c/src/udprelay.cpp#L495) logic for receiving FIBRE block announcements. The way that FIBRE nodes announce to "default" nodes is the same as the way that "default" nodes announce to each other. [Below](https://github.com/w0xlt/bitcoinfibre/blob/45897a826b37eb417cba93ac17ff7bebb272893c/src/net_processing.cpp#L881) the first snippet you linked, is this `if` statement:

https://github.com/bitcoin/bitcoin/blob/a45d53cab556505048c387429fd07188e4c40c3d/src/net_processing.cpp#L1659-L1660

which is essentially identical to the logic used today, except `fPreferHeadersAndIDs->m_requested_hb_compactblocks` (https://github.com/bitcoin/bitcoin/commit/3b6bfbce386f61dcbb366f08cfff55c3882f429c)

**Comment by Crypt-iQ** [comment]
> > The relevant code is in UDPRelayBlock
> 
> `UDPRelayBlock` is relevant only for FIBRE to FIBRE connections, and these also have [bespoke](https://github.com/w0xlt/bitcoinfibre/blob/45897a826b37eb417cba93ac17ff7bebb272893c/src/udprelay.cpp#L495) logic for receiving FIBRE block announcements. The way that FIBRE nodes announce to "default" nodes is the same as the way that "default" nodes announce to each other. [Below](https://github.com/w0xlt/bitcoinfibre/blob/45897a826b37eb417cba93ac17ff7bebb272893c/src/net_processing.cpp#L881) the first snippet you linked, is this `if` statement:
> 
> [bitcoin/src/net_processing.cpp](https://github.com/bitcoin/bitcoin/blob/a45d53cab556505048c387429fd07188e4c40c3d/src/net_processing.cpp#L1659-L1660)
> 
> Lines 1659 to 1660 in [a45d53c](/bitcoin/bitcoin/commit/a45d53cab556505048c387429fd07188e4c40c3d)
>  if (state.fPreferHeaderAndIDs && (!fWitnessEnabled || state.fWantsCmpctWitness) && 
>          !PeerHasHeader(&state, pindex) && PeerHasHeader(&state, pindex->pprev)) { 
> 
> which is essentially identical to the logic used today, except `fPreferHeadersAndIDs->m_requested_hb_compactblocks` ([3b6bfbc](https://github.com/bitcoin/bitcoin/commit/3b6bfbce386f61dcbb366f08cfff55c3882f429c))

Gotcha, my bad for the misunderstanding. That means my comment about FIBRE is wrong in your open PR.

**Comment by ajtowns** [comment]
> After [#27675](https://github.com/bitcoin/bitcoin/pull/27675) (but also before, with a few more restrictions), we relay any tx from our mempool (except for those that arrived after the last INV sent to the peer), so someone wanting to fingerprint our mempool could just ask for those transactions directly instead I think and then check whether we send it to them or answer with `NOTFOUND`.

I don't think this issue makes sense as an attack -- we aim to avoid giving away tight timing information about exactly when txs entered our mempool, but not whether txs are in our mempool at all. But this attack can't be timed tightly -- you need ten minutes' worth of proof of work to perform the attack.

One possible way to mitigate this attack nearly completely according to this model would probably be to update all peers' `m_last_inv_sequence` when a new block header is seen, and, if a GETBLOCKTXN request is made, to also request any txs from the mempool whose sequence exceeded that value.
